### PR TITLE
Initial iteration on EPIC implementation

### PR DESCRIPTION
**Pull request by Rocamonde**

### CODE REVIEW

File: src/epic/samplers.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+
+class DummyGymStateSampler(BaseSampler[Tuple[npt.NDArray[bool], npt.NDArray]], BaseGymSampler):
+    def sample(self):
+        state_sample = np.array([self.space.sample() for _ in range(self.n_samples)])
+        done_sample = np.zeros(self.n_samples, dtype=np.bool_)
+        return done_sample, state_sample
+
+
+class PreloadedDataSampler(BaseSampler[np.ndarray]):
```
Comment (levmckinney) [inline src/epic/samplers.py:41]:
```suggestion
class PreloadedDataSampler(BaseSampler[npt.NDArray]):
```

File: src/epic/samplers.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    """A sampler that samples from a preloaded dataset."""
+    data: np.ndarray
+    n_samples: Optional[int]
+    rng: np.random.Generator
+    def __init__(self, data: np.ndarray, n_samples: Optional[int] = None, rng: Optional[np.random.Generator] = None):
```
Comment (levmckinney) [inline src/epic/samplers.py:46]:
```suggestion
    def __init__(self, data: npt.NDArray, n_samples: Optional[int] = None, rng: Optional[np.random.Generator] = None):
```

File: src/epic/samplers.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        if self.n_samples is not None and self.data.shape[0] < self.n_samples:
+            raise ValueError(f"n_samples ({self.n_samples}) must be less than the number of data points ({self.data.shape[0]})")
+
+    def sample(self):
+        return self.data[self.rng.integers(0, self.data.shape[0], self.n_samples)] \
+                if self.n_samples is not None \
+                else self.data
+
+
+def make_sampler(fn: Callable[[], np.ndarray]) -> BaseSampler[np.ndarray]:
```
Comment (levmckinney) [inline src/epic/samplers.py:72]:
```suggestion
def make_sampler(fn: Callable[[], npt.NDArray]) -> BaseSampler[np.ndarray]:
```

File: src/epic/samplers.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    """Decorator to create a sampler from a function.
+
+    The function should take no arguments and return a numpy array.
+
+    Args:
+        fn: The function to use to create the sampler.
+
+    Returns: A sampler instance.
+    """
+    class FunctionSampler(BaseSampler[np.ndarray]):
```
Comment (levmckinney) [inline src/epic/samplers.py:82]:
```suggestion
    class FunctionSampler(BaseSampler[npt.NDArray]):
```

File: src/epic/samplers.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        def sample(self) -> np.ndarray:
```
Comment (levmckinney) [inline src/epic/samplers.py:83]:
```suggestion
        def sample(self) -> npt.NDArray:
```

File: src/epic/utils.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    function: Fn, *grouped_arguments: Dict[str, np.ndarray]
+) -> np.ndarray:
+    arg_names = [
+        arg_name
+        for arg_group in grouped_arguments
+        for arg_name in list(arg_group.keys())
+    ]
+    assert len(arg_names) == len(
+        set(arg_names)
+    ), "Argument names must not be unique within and across groups"
```
Comment (levmckinney) [inline src/epic/utils.py:53]: Must **not** be unique?

File: src/epic/utils.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        batch_lengths.append(shapes[0])
+
+    class Info(TypedDict):
+        axis: int
+        array: np.ndarray
+
+    map_arg_to_info: Dict[str, Info] = dict()
+    for index, arg_group in enumerate(grouped_arguments):
+        for arg_name, arg_val in arg_group.items():
+            map_arg_to_info[arg_name] = {"axis": index, "array": arg_val}
```
Comment (levmckinney) [inline src/epic/utils.py:70]: Did you mean to use the `Info` class here?

File: src/epic/utils.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        # this is how the shape of the array should be:
+        assert tiled_array.shape == tuple((*batch_lengths, *array.shape[1:]))
+        arguments[arg_name] = tiled_array
+
+    return multidim_batch_call(
+        function, arguments=arguments, num_batch_dims=len(batch_lengths)
+    )
+
+
+def product_batch_wrapper(function: Fn, nested=True):
```
Comment (levmckinney) [inline src/epic/utils.py:95]: A doc string here would be good. Right now it's quite hard to understand what this code is doing.

File: src/epic/types.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        Args:
+            state: Current states of shape `(batch_size,) + state_shape`.
+            action: Actions of shape `(batch_size,) + action_shape`.
+            next_state: Successor states of shape `(batch_size,) + state_shape`.
+        Returns:
+            Computed rewards of shape `(batch_size,`).
+        """
+
+
+class CoverageDistribution(Protocol):
```
Comment (levmckinney) [inline src/epic/types.py:32]: This does not seem to be used anywhere.

File: tests/test_epic_dist.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
+import numpy as np
```
Comment (levmckinney) [inline tests/test_epic_dist.py:1]: We could use `hypothesis` to make these tests more robust. You can checkout my tests in https://github.com/HumanCompatibleAI/tabular-baseline/blob/4d32afd3ba30f6eb50e9203735479a2a725e96df/tests/test_epic_distance.py

File: tests/test_epic_dist.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    dist = epic.EPIC(
+        state_sampler=DummyGymStateSampler(space=state_space, n_samples=100),
+        action_sampler=GymSpaceSampler(space=action_space, n_samples=100),
+        discount_factor=1,
+    ).distance(x, y)
+
+    assert isinstance(dist, float)
+
+
+def test_epic_dist_reward_equivalence():
```
Comment (levmckinney) [inline tests/test_epic_dist.py:36]: We should have a negative version of this test.

**Review by levmckinney — APPROVED** [review]
This is a beautiful implementation. I have a few foibles about testing and documentation but overall great work!

File: src/epic/distances/base.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        /,
+        n_samples_can: Optional[int],
+    ) -> types.RewardFunction:
+        """Canonicalize a reward function.
+
+        Args:
+            reward_function: The reward function to canonicalize.
+            n_samples_cov: The number of samples to draw from the coverage distribution.
+            n_samples_can: The number of samples to draw for the canonicalization step
+                for each sample of the coverage distribution. The total number of
```
Comment (ejnnr) [inline src/epic/distances/base.py:59]: Note that for some canonicalization procedures other than EPIC, it makes more sense to use only one set of canonicalization samples. E.g. for divergence-free rewards, we'll try fitting a small neural network during the canonicalization step, and I think we'll only want to do this once and then use the same network for all coverage samples (both for efficiency reasons and to make sure we're actually shaping with a valid potential).

Not sure what that should mean for parameters. We could just use `n_samples_cov * n_samples_can` samples for div-free rewards, that would make the total number of samples consistent. Or we could use `n_samples_can`, that would make the number of samples used to compute the canonicalization consistent.

File: src/epic/distances/base.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+            y: The second reward function.
+            n_samples_cov: The number of samples to draw from the coverage distribution.
+            n_samples_can: The number of samples to draw for the canonicalization step
+                for each sample of the coverage distribution. The total number of
+                samples drawn is ``n_samples_cov * n_samples_can``.
+
+        Returns:
+            The distance between the two reward functions.
+        """
+        x_canonical = self.canonicalize(x, n_samples_can)
```
Comment (ejnnr) [inline src/epic/distances/base.py:107]: Related to my comment above: this will mean that divergence-free rewards will only use `n_samples_cov + n_samples_can` total samples, since the canonicalized reward function won't sample anything itself every time it's called. Maybe that's what we want, just noting the difference (e.g. div-free rewards will probably need a higher `n_samples_can`)

File: src/epic/distances/base.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    form.
+    """
+
+    state_sampler: samplers.BaseSampler[samplers.StateSample]
+    action_sampler: samplers.BaseSampler[npt.NDArray]
+    coverage_sampler: samplers.BaseSampler[samplers.CoverageSample]
+
+    def __init__(
+        self,
+        state_sampler: samplers.BaseSampler[samplers.StateSample],
```
Comment (ejnnr) [inline src/epic/distances/base.py:36]: Note that having a distinct state and action sampler is distinct to EPIC (neither DARD, nor divergence-free and outflow-free rewards have that)

File: src/epic/distances/epic.py [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+            n_samples_cov = state.shape[0]
+
+            assert n_samples_cov == action.shape[0] == next_state.shape[0] == done.shape[0]
+
+            n_samples = n_samples_cov * n_samples_can
+
+            # Copy each sample in n_samples_cov to n_samples_can times.
+            state, action, next_state, done = utils.broadcast(state, action, next_state, done, n_samples_can)
+
+            # Create n_samples_cov * n_samples_can samples.
```
Comment (ejnnr) [inline src/epic/distances/epic.py:84]: FWIW, it's not obvious to me this is better than using all the canonicalization samples for each coverage sample. I'd imagine if the cost of sampling is low compared to computing the means, the current version makes sense, but if sampling is expensive, you'd rather just use all the samples each time?

File: src/epic/distances/epic.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+    of reward functions. Instead, instantiate the EPIC class and call the
+    ``distance`` method on the instance multiple times.
+
+    Args:
+        x: The first reward function.
+        y: The second reward function.
+        state_sampler: A sampler for the state space.
+        action_sampler: A sampler for the action space.
+        discount_factor: The discount factor of the MDP.
+        nested: Whether sampling is nested over any expectation operators. See
```
Comment (ejnnr) [inline src/epic/distances/epic.py:166]: Looks like the doc string doesn't match the arguments here

File: tests/test_epic_dist.py [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+def rew_fn_1(state, action, next_state, _):
+    return state + next_state + action
+
+
+def rew_fn_2(state, action, next_state, _):
+    return state**2 + next_state**2 + 2 * action
+
+
+def rew_fn_1_potential_shaping(state, action, next_state, _):
```
Comment (ejnnr) [inline tests/test_epic_dist.py:17]: Might end up being useful to have a wrapper that takes a `states -> floats` function and returns a `RewardFunction`, then we could easily test things like this for a bunch of combinations of reward functions and potentials. (Not super high priority)

If the wrapper was written such that it also supported pytorch tensors and backprop, I think it could also be used for the implementation of div-free rewards (where we have a neural net describing the `states -> floats` function, but care about the associated reward function for canonicalization)

File: src/epic/distances/epic.py [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+            return (
+                reward_function(state, action, next_state, done)
+                + self.discount_factor * term_1
+                - term_2
+                - self.discount_factor * term_3
+            )
+
+        return canonical_reward_fn
+
+    def _distance(
```
Comment (ejnnr) [inline src/epic/distances/epic.py:117]: divergence-free rewards will use the same `_distance` implementation, so we'll need to move this somewhere reusable

**Review by ejnnr — COMMENTED** [review]
Left some comments, mainly regarding what might need to be changed for the upcoming implementation of divergence-free rewards. Not clear those changes should be made in this PR though.

Didn't spot any bugs in the implementation, but not very confident (I didn't carefully follow all of the reshaping etc.)

To be really sure the implementation is correct, you could check against known distances in specific cases (e.g. the gridworld ones from the EPIC paper?). The invariance test is useful, but when I was implementing divergence-free rewards in a tabular setting, I passed the invariance test as well as several other consistency checks, and the implementation was still incorrect.

File: src/epic/distances/base.py (hunk 14 above)
Comment (anshradh) [reply src/epic/distances/base.py:36]: Since it seems like the coverage_sampler will be needed for both, but the state and action samplers are unique to EPIC, they should both be optional here in the abc

File: src/epic/distances/base.py (hunk 12 above)
Comment (anshradh) [reply src/epic/distances/base.py:59]: (Discussed this with Erik in person as well)

I think having these parameters is fine, we should probably just specify in the subclasses how they're used differently and that EPIC uses `n_samples_cov * n_samples_can` total samples while e.g. div-free rewards uses `n_samples_cov + n_samples_can` total samples. 

I think the function definition is also missing n_samples_cov as an argument?

File: tests/test_epic_dist.py (hunk 17 above)
Comment (anshradh) [reply tests/test_epic_dist.py:17]: Just to make sure I'm following - this just adds the `states -> floats` function as a potential to a `RewardFunction`?

I think implementing this wrapper and making it support backprop seems good. I could also potentially take this on when I implement the div-free reward distance calculation.

Comment (ejnnr) [reply tests/test_epic_dist.py:17]: Yeah exactly, that's the wrapper I had in mind. The only issue with supporting backprop (AFAICT) would be that the `RewardFunction` type currently requires numpy arrays. I'd guess we don't just want to generalize `RewardFunction` to allow pytorch tensors because presumably not all of the code supports pytorch (though I haven't checked). Maybe just have a more general type signature specifically for that wrapper

Comment (ejnnr) [reply tests/test_epic_dist.py:17]: Oh, to be clear: I think the wrapper doesn't necessarily _add_ to a reward function that it takes as input---it could just return the part that would need to be added, i.e. gamma * Phi(s') - Phi(s), which is itself a valid reward function. If needed, this "pure potential shaping reward function" could then be added to some other existing reward function outside the wrapper. Though tbf, in most use cases we probably _do_ want to add this pure potential shaping term to a specific reward function, so writing the wrapper that way also seems reasonable

File: src/epic/distances/epic.py (hunk 18 above)
Comment (anshradh) [reply src/epic/distances/epic.py:117]: I'd be in favor of a `PearsonDistanceMixin` that both `EPIC` and `DivergenceFree` inherit from.

Comment (ejnnr) [reply src/epic/distances/epic.py:117]: Sounds like a good solution to me!

**Comment by Rocamonde** [comment]
Hi @ejnnr @anshradh thank you both for your comments! It would be nice to have a discussion about this over a call. I also tried to think about how to make the implementation as general as possible for future distance metrics. My goal is that this could once become the go-to package to measure distances between rew fns for people doing reward learning, with a range of distance metrics, integrations, and robust sampling algorithms with convergence guarantees.

**Comment by anshradh** [comment]
Yeah I should have mentioned that this was a very clean and easy to follow implementation, so hats off to you @Rocamonde! 

I'm going to be working on extending this for divergence-free reward distances, so I think most of mine and Erik's comments were towards that aim. Honestly the PR mostly lgtm already and I can also add/refactor things as needed as I do my implementation (which I've started on a branch based off of this current one), so it may just be better to leave the PR mostly as is and for me to make the modifications as needed. Happy to talk more about this some time though!
